### PR TITLE
Support ranged upgrades

### LINKED ISSUE #392: Passing a version range to `-P` results in no updates
It seems that passing a package range to `-P` causes it `pip-compile-multi` to not upgrade the package in question at all, even when upgrades are available.

This is reproducible with:
```
# requirements/base.in
markupsafe
```
```
# requirements/base.txt
markupsafe==2.0.1
    # via -r requirements/base.in
```
``` shell
$ pip-compile-multi  -P 'markupsafe<2.1.2'
Locking requirements/prod.in to requirements/prod.txt. References: []
```
Observing that while markupsafe has plenty of versions between 2.0.1 and 2.1.2, none are upgraded to.

Am I misunderstanding how `-P` is expected to work? I had thought I'd seen this work sometimes, but now can't actually reproduce that. `-P` seems to be documented to work the same as the equivalent `pip-tools` option, which does accept ranges and works with them.

### LINKED ISSUE #393: Using `--autoresolve` with `-P` and packages with version ranges fails confusingly
It seems that passing a package range to `-P` when also using `--autoresolve` and more than one layer of requirements causes `pip-compile-multi` to get confused about where some of the version pins are coming from.

This is reproducible with:

`requirements/prod.in`:
```
markupsafe
```
`requirements/prod.txt`:
```
# SHA1:2f781f20c9698050fd9afac72c0364db5deacb05
#
# This file is autogenerated by pip-compile-multi
# To update, run:
#
#    pip-compile-multi
#
markupsafe==2.0.1
    # via
    #   -c requirements/tests.txt
    #   -r requirements/prod.in
```
`requirements/tests.in`:
```
-r prod.in
```
`requirements/tests.txt`:
```
# SHA1:939d3916ae2e78296c77d899249472d04707c772
#
# This file is autogenerated by pip-compile-multi
# To update, run:
#
#    pip-compile-multi
#
-r prod.txt
```

then:
``` shell
$ pip-compile-multi --autoresolve -P 'markupsafe<2.1.2'
Creating a temporary file with all dependencies at requirements/tests.txt
Locking requirements/prod.in to requirements/prod.txt. References: []
Locking requirements/tests.in to requirements/tests.txt. References: ['requirements/prod.in']
Package markupsafe was resolved to different versions in different environments: 2.1.1 and 2.0.1
Traceback (most recent call last):
  File "/tmp/venv/lib/python3.10/site-packages/pipcompilemulti/cli_v1.py", line 26, in cli
    recompile()
  File "/tmp/venv/lib/python3.10/site-packages/pipcompilemulti/actions.py", line 31, in recompile
    compile_topologically(env_confs, deduplicator)
  File "/tmp/venv/lib/python3.10/site-packages/pipcompilemulti/actions.py", line 38, in compile_topologically
    if env.maybe_create_lockfile():
  File "/tmp/venv/lib/python3.10/site-packages/pipcompilemulti/environment.py", line 51, in maybe_create_lockfile
    self.create_lockfile()
  File "/tmp/venv/lib/python3.10/site-packages/pipcompilemulti/environment.py", line 72, in create_lockfile
    self.fix_lockfile()
  File "/tmp/venv/lib/python3.10/site-packages/pipcompilemulti/environment.py", line 134, in fix_lockfile
    sections = [
  File "/tmp/venv/lib/python3.10/site-packages/pipcompilemulti/environment.py", line 135, in <listcomp>
    self.fix_pin(section)
  File "/tmp/venv/lib/python3.10/site-packages/pipcompilemulti/environment.py", line 211, in fix_pin
    raise RuntimeError(
RuntimeError: Please add constraints for the package version listed above
```

### PR DESCRIPTION
**Pull request by PeterJCLaw**
This introduces a generic mechanism for adding test cases based on input files, then uses that to validate the failure mode reported in https://github.com/peterdemin/pip-compile-multi/issues/392. Finally this applies a fix which then passes the added test.

Fixes #392

This also turns out to have fixed https://github.com/peterdemin/pip-compile-multi/issues/393, so I've added a test for that too.

Fixes #393

### CODE REVIEW

File: pipcompilemulti/features/upgrade.py [hunk 1]
```diff
@@ -78,10 +80,19 @@ def reset(self):
...
+    def package_specs(self):
+        """List of package specs to upgrade."""
+        return self.value or []
+
     @property
     def package_names(self):
         """List of package names to upgrade."""
-        return self.value or []
+        def name_from_spec(name):
+            match = re.match(r'(?P<package>[a-z0-9-_.]+)', name)
```
Comment (peterdemin) [inline pipcompilemulti/features/upgrade.py:92]: Let's move regex compilation to a class attribute, like done here: https://github.com/peterdemin/pip-compile-multi/blob/master/pipcompilemulti/dependency.py#L50

File: pipcompilemulti/features/upgrade.py [hunk 2]
```diff
@@ -78,10 +80,19 @@ def reset(self):
...
+            assert match is not None
```
Comment (peterdemin) [inline pipcompilemulti/features/upgrade.py:93]: I'm not a fan of having `assert`s outside of tests. The following line will produce an exception anyway.

**Review by peterdemin — APPROVED** [review]
Thank you so much! The code is of high quality. I left a couple of minor comments.

File: pipcompilemulti/features/upgrade.py (hunk 2 above)
Comment (PeterJCLaw) [reply pipcompilemulti/features/upgrade.py:93]: :+1: Alternatively, is there a pattern I can follow to produce a nice error message for the user? Ideally something which helps them understand and likely fix the issue (i.e: that the package name specified likely isn't a valid package).

File: pipcompilemulti/features/upgrade.py (hunk 1 above)
Comment (PeterJCLaw) [reply pipcompilemulti/features/upgrade.py]: 3b471d8bd0013db7b845f71dfe0de6d8fef2a1b5

File: pipcompilemulti/features/upgrade.py (hunk 2 above)
Comment (PeterJCLaw) [reply pipcompilemulti/features/upgrade.py]: d4ff1ad362c3954027db37d8554e14c0392affe6

**Review by peterdemin — APPROVED** [review]

File: tests/test_cli_v1.py [hunk 3]
```diff
@@ -76,7 +78,11 @@ def test_package_upgrade(name, args):
...
         actual_root = Path(tmp_dir)
-        expected = _load_tree(expected_root)
+        expected = _load_tree(
+            expected_root,
+            # Cope with posix style reference data on Windows
+            replace_name=str(expected_root).replace(os.path.sep, '/'),
+        )
         actual = _load_tree(actual_root)
 
         assert actual == expected
```
Comment (peterdemin) [inline tests/test_cli_v1.py:88]: Path separators in the output are platform-specific too. I think it's safe for testing purposes to normalize all `\` to `/` in the `actual` output.

File: tests/test_cli_v1.py [hunk 4]
```diff
@@ -41,3 +46,43 @@ def test_v1_verify_exits_with_zero():
     runner = CliRunner()
     result = runner.invoke(cli, ['verify'])
     assert result.exit_code == 0
+
+
+def _load_tree(root, replace_name=None):
+    if not replace_name:
+        replace_name = str(root)
+    return {
+        x.relative_to(root).name: x.read_text().replace(replace_name, 'ROOT').replace('\\', '/')
```
Comment (peterdemin) [inline tests/test_cli_v1.py:55]: I added path separator normalization ☝🏻

**Comment by peterdemin** [comment]
Thank you!

**Comment by peterdemin** [comment]
Released in [2.6.3](https://pypi.org/project/pip-compile-multi/2.6.3/)

File: tests/test_cli_v1.py (hunk 4 above)
Comment (PeterJCLaw) [reply tests/test_cli_v1.py:55]: Oh, right, I missed the separators outside the root :facepalm:, :+1:

**Pull request merged**